### ISSUE TITLE
Pull replication not working when device goes offline and back online

### ISSUE THREAD

**Issue by sethrosetter** [couchbase/couchbase-lite-java-core#976]
Steps to reproduce using CBL-Android 1.2.0/043e0eedcf22febf1dcfe2644ac242019da3daf1):

Nexus 5 - Android 4.4.4 device
Nexus 5 - API 23 x86 emulator

- Deploy grocery-sync app to device and emulator
- Ensure the are connected with an sync_gateway instance (Couchbase Sync Gateway/1.2.0(79;9df63a5))
- Turn the Wifi off on the device
- Create ~8 items on emulator (emulator is still online, and pushes docs to sync_gateway)
- Turn the Wifi on on the device

Expected:
Docs should be pulled to device
Actual:
Docs do not sync - observed the following error in log cat - https://gist.github.com/sethrosetter/442f1b167e1c619889b6#file-gistfile1-txt-L88-L112

**Comment by zgramana** [comment]
Notes:

The gist confirms a deadlock, as the call to [batcher.waitForPendingFutures](https://gist.github.com/sethrosetter/442f1b167e1c619889b6#file-gistfile1-txt-L148) never completes.

Note: on ARM processors [this block of code](https://github.com/couchbase/couchbase-lite-java-core/blob/3a1808df6525e9f1454822df441cd9eef69b1550/src/main/java/com/couchbase/lite/support/Batcher.java#L226-L229) is not garaunteed safe:

```java
        if (scheduled && delay < scheduledDelay)
            unschedule();
        if (!scheduled) {
            scheduled = true;
```

Because of ARM's instruction re-ordering mechanism, the `scheduled = true` instruction may get re-ordered to execute before the `if (!scheduled)` check! In an old implementation of our custom work executor in .NET, I had to [use a memory barrier](https://github.com/couchbase/couchbase-lite-net/blob/53726231685df3650b92f9e39f0464f6a11d3177/src/Couchbase.Lite.Shared/Util/SingleTaskThreadpoolScheduler.cs#L50-L54) to prevent this from happening. @borrrden smartly refactored away the read-then-set pattern which removed the need for the memory barrier call.

**Comment by pasin** [comment]
It's hard to tell from the gist where the deadlock is. After calling waitForPendingFutures, ChangeTracker was still receiving changes and Puller was still adding changes to the inbox which is a batcher.

**Comment by zgramana** [comment]
I noticed [this log statement](https://github.com/couchbase/couchbase-lite-java-core/blob/d704e12d973a1dcd763dd897ad34f0fee3950b5f/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L1024) never shows up in the gist: 
```java
Log.d(TAG, "[waitForPendingFutures()] END - thread id: " + Thread.currentThread().getId());
```

**Comment by zgramana** [comment]
It also look like it never reaches [this log statement](https://github.com/couchbase/couchbase-lite-java-core/blob/d704e12d973a1dcd763dd897ad34f0fee3950b5f/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L1049):
```java
                batcher.waitForPendingFutures();
            }

            // wait for pending featurs completed
            Log.d(TAG, "waitPendingFuturesCompleted()");
```

**Comment by zgramana** [comment]
It might not hit [this infinite loops exit criteria](https://github.com/couchbase/couchbase-lite-java-core/blob/d704e12d973a1dcd763dd897ad34f0fee3950b5f/src/main/java/com/couchbase/lite/support/Batcher.java#L239-L240) under two conditions:
```java
            synchronized (mutex) {
                if (inbox.isEmpty())
                    break;
            }
```
1. Can't acquire the mutex
2. Inbox isn't empty as the time this is called.

**Comment by hideki** [comment]
Without Batcher Log, it is hard to identify which logic causes dead-lock.

**Comment by zgramana** [comment]
Since thread A's call to `future.get()` is a blocking call on thread B, is there a chance that the thread B is blocking on thread A?

This seems like [a decent candidate](https://github.com/couchbase/couchbase-lite-java-core/blob/d704e12d973a1dcd763dd897ad34f0fee3950b5f/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L974), since it gets called via `PullerInternal.stop()` which [gets called by `restart`](https://github.com/couchbase/couchbase-lite-java-core/blob/85c93168b515dbde1116f41d373972b6b8209908/src/main/java/com/couchbase/lite/replicator/Replication.java#L226):
```java
            // Not waiting here to avoid deadlock as this method is executed
            // inside the same WorkExecutor as the downloadsToInsert batcher.
            // All scheduled objects will be waited to completed by calling
            // waitForAllTasksCompleted() below.
            downloadsToInsert.flushAll(false);
```

**Comment by sethrosetter** [comment]
sync_gateway logs around the time of this changes request  - https://gist.github.com/sethrosetter/9be61a8b8b7795d77940#file-sync_gateway_error-log-L1344

**Comment by zgramana** [comment]
Is it possible [this commit](https://github.com/couchbase/couchbase-lite-java-core/commit/9468652535f2b273b76859e5a63b2267d9c2edea) introduced a subtle issue, or perhaps didn't fully resolve the original issue?

> **Fixed dead lock when calling Batcher.flushAll()**
>
> PullerInternal calls Batcher.flushAll() from the same work executor given to the Batcher. Now make Batcher.flushAll() with an option to wait or not to wait for all objects to be processed.
> 
> Also make sure that while executing flushAll(), all new queued objects will be scheduled to work by flushAll().
>

**Comment by zgramana** [comment]
Just a breadcrumb trail from a chat with @pasin, [found an incorrect lock object](https://github.com/couchbase/couchbase-lite-java-core/blob/d704e12d973a1dcd763dd897ad34f0fee3950b5f/src/main/java/com/couchbase/lite/support/Batcher.java#L171-L173) (should use `mutex`):
```java
                        synchronized (Batcher.this) {
                            lastProcessedTime = System.currentTimeMillis();
                        }
```

**Comment by pasin** [comment]
One thing from the log messages is that the puller replicator keeps being in not-running state but change tracker is active. It is possible that when bring the WiFi offline (app might be in the background when doing that), the pull replicator might be in stopped state and never woke up.

**Comment by hideki** [comment]
I agree with @pasin. From log message, it seems pull replicator is not running. Then it can not consume changes event that ChangeTracker retrieved from Sync Gateway.

**Comment by pasin** [comment]
Maybe when trying to reproduce the issue, we should watch to see the replication stops running before changing the wifi on again.

**Comment by pasin** [comment]
@sethrosetter I'm wondering if this action from the log file is from the pull replicator or from you adding/updating an item after turning the WiFi on.

https://gist.github.com/sethrosetter/442f1b167e1c619889b6#file-gistfile1-txt-L64-L67

**Comment by sethrosetter** [comment]
@pasin I'm not totally sure. But I think I did add an item to see if that triggered the replication to wake up. Not 100% sure but I think I did.

**Comment by pasin** [comment]
I run the GrocerySync release/1.2.0-sqlite branch (using the same sync-gateway url that Seth uses):
1. Run GrocerySync on an emulator, created 5 items. All items got pushed.
2. On my device, I have the airplane mode with WiFi off.
3. I run the app from Android Studio, and checking the log. I wait until I see the replicator stopped from the log file due to an error.
4. I turned WiFi on and then I saw the replication went online. Pull Replicator tried to sync but there was a NPE happens due to db = null so sync didn't success. I think db was cleared when the replicator is stopped but never get set back after that.

Here is the full log: https://gist.github.com/pasin/6e434157458d0aaa4538

**Comment by hideki** [comment]
From log @pasin provided, I guess following things occurs:
1. Device becames offline (IDLE/RUNNING -> OFFLINE)
2. ChangeTracker and/or on-going pull replication fails with HTTP error (This could be SocketTimeoutException). 
3. Error of No 2, makes pull replicator stops. (OFFLINE -> STOPPING -> STOPPED)
4. Device becomes online
5. No 4 leads to start ChangeTracker. 
6. Replicator state can not change from STOPPED to RUNNING
7 As result, ChangeTracker is running, but pull replicator keeps to stop.

Areas to look into:
1. How replicator handles Device offline notification (goOffline)
2. How replicator/ChangeTracker handles error during OFFLINE state
3. How replicator handles Device online notification (goOnline)

**Comment by hideki** [comment]
Some reason this ticket was automatically closed twice?????

**Comment by hideki** [comment]
Hi @pasin,
I tried to reproduce this with following your steps for more than 10 times. With my testing, the replicator has never stopped during offline state. Are there any technique to reproduce this?

**Comment by pasin** [comment]
@hideki, have you tried to reproduce with the AWS demo URL? Also, step 2 and 3 are important to reproduce the issue.

**Comment by hideki** [comment]
@pasin,

URL that @sethrosetter used was no longer available. So I used demo `demo-mobile.couchbase.com` server.

> 3. I ran the app from the Android Studio, and waited until I saw the replicator stopped from the log messages.

I can not observe that the replicator stops without any action.

I observed replication stops in following scenarios:
- I observed replicator stops by `Manager.close()` when app went to background to disable/enable wifi.
- I also observed replication stops by network error during `startReplication` which was triggered by `goOnline`.

**Comment by pasin** [comment]
From the log that I posted, there is an transient error (UnknownHostException) happens while getting remote checkpoint after starting the replicator. Then there was a retry logic happening there until `retry exhausted` which led the replication to stopped.

```
02-04 14:52:44.633 30636-30756/com.couchbase.grocerysync V/Sync: com.couchbase.lite.support.RemoteRequest@31db55: RemoteRequest finally block.  url: http://ec2-54-197-148-230.compute-1.amazonaws.com:4985/db/_local/f963aba90e0ea60c097ef8af75c0682d96a23965
02-04 14:52:44.633 30636-30756/com.couchbase.grocerysync V/Sync: com.couchbase.lite.support.RemoteRequest@31db55: RemoteRequest calling respondWithResult.  url: http://ec2-54-197-148-230.compute-1.amazonaws.com:4985/db/_local/f963aba90e0ea60c097ef8af75c0682d96a23965, error: java.net.UnknownHostException: Unable to resolve host "ec2-54-197-148-230.compute-1.amazonaws.com": No address associated with hostname
02-04 14:52:44.633 30636-30756/com.couchbase.grocerysync D/Sync: com.couchbase.lite.support.RemoteRequestRetry$1@2a8b17c: RemoteRequestRetry inner request finished, url: http://ec2-54-197-148-230.compute-1.amazonaws.com:4985/db/_local/f963aba90e0ea60c097ef8af75c0682d96a23965
02-04 14:52:44.633 30636-30756/com.couchbase.grocerysync D/Sync: com.couchbase.lite.support.RemoteRequestRetry@cbd977: isTransientError called, httpResponse: null e: java.net.UnknownHostException: Unable to resolve host "ec2-54-197-148-230.compute-1.amazonaws.com": No address associated with hostname
02-04 14:52:44.633 30636-30667/com.couchbase.grocerysync W/Sync: com.couchbase.lite.replicator.ReplicationInternal$10@95c3aa4: error getting remote checkpoint
02-04 14:52:44.634 30636-30756/com.couchbase.grocerysync D/Sync: com.couchbase.lite.support.RemoteRequestRetry@cbd977: its an ioexception, return true
02-04 14:52:44.634 30636-30756/com.couchbase.grocerysync D/Sync: com.couchbase.lite.support.RemoteRequestRetry$1@2a8b17c: RemoteRequestRetry failed, but transient error.  retries exhausted. url: http://ec2-54-197-148-230.compute-1.amazonaws.com:4985/db/_local/f963aba90e0ea60c097ef8af75c0682d96a23965
02-04 14:52:44.634 30636-30756/com.couchbase.grocerysync W/Sync: com.couchbase.lite.replicator.ReplicationInternal$10@4fb200d: error getting remote checkpoint
02-04 14:52:44.634 30636-30667/com.couchbase.grocerysync E/Sync: com.couchbase.lite.replicator.PullerInternal@8f6ee9b: Progress: set error = java.net.UnknownHostException: Unable to resolve host "ec2-54-197-148-230.compute-1.amazonaws.com": No address associated with hostname
02-04 14:52:44.635 30636-30756/com.couchbase.grocerysync E/Sync: com.couchbase.lite.replicator.PusherInternal@f9a9613: Progress: set error = java.net.UnknownHostException: Unable to resolve host "ec2-54-197-148-230.compute-1.amazonaws.com": No address associated with hostname
02-04 14:52:44.635 30636-30666/com.couchbase.grocerysync D/GrocerySync: Replication : com.couchbase.lite.replicator.Replication@a6af6fa changed.
02-04 14:52:44.635 30636-30666/com.couchbase.grocerysync D/GrocerySync: Replicator processed 0 / 0
02-04 14:52:44.635 30636-30756/com.couchbase.grocerysync D/Sync: [fireTrigger()] => STOP_GRACEFUL
```

**Comment by hideki** [comment]
I observed same error that was triggered by `goOnline` -> `startReplicator()` -> `UnknownHostException `. Do you remember which timing you observed this? after `goOnline()`?

**Comment by hideki** [comment]
`UnknownHostException` occurs in case state is `OFFLINE`. In case of `OFFLINE` state, Replicator should not try any network activities. Also requires different network error handling during `OFFLINE` state.

**Comment by pasin** [comment]
1. The `UnknownHostException ` started from this line: https://gist.github.com/pasin/6e434157458d0aaa4538#file-gistfile1-txt-L179. Then, there were a bunch of retires and then replicator stopped. 

2. After the WiFi is On,  the reachability tried to bring the replicator online:
https://gist.github.com/pasin/6e434157458d0aaa4538#file-gistfile1-txt-L279-L283. However, when the replicator became Online, the database was already cleared and led to the NPE exception.

3. When the replicators were brought Online by the reachability, from the log, I couldn't tell if the pusher was actually online or not. Also I couldn't tell if the process to bring the replicator online is correct or not.

**Comment by hideki** [comment]
This is duplicate ticket with #1026.  Fix is associated with #1026. I will close this.
